### DISCUSSION TITLE
[patrol_log] Error in VS Code terminal  Unhandled exception: FormatException: Missing extension byte at offset

### DISCUSSION THREAD

**Discussion by EvgenSilkov** [leancodepl/patrol#2455] [Q&A] [answered]
On patrol 3.12.0 patrol_cli 3.3.0  it worked like a charm.
patrol_cli 3.4.1 with smiles in console crashes semi-randomly midrun on different tests while trying to put test result to the console. Tests keep running, but console supervisor is shut down.
Tests are quite different, message, though, is the same (provided below)
I can't pinpoint it to a specific character in test. When trying to run smaller amount of tests it passes without a hitch.
I can't reproduce it consistently so I'm not bothering developers with filling issue.

UPD patrol_log works bad with non ASCII characters  in VS Code native terminal.
 switched to Ghostty terminal and it happens not as often.

Log:

```
Unhandled exception:
FormatException: Missing extension byte (at offset 1066)
#0      _Utf8Decoder.convertChunked (dart:convert-patch/convert_patch.dart:1956:7)
#1      _Utf8ConversionSink.addSlice (dart:convert/string_conversion.dart:304:28)
#2      _Utf8ConversionSink.add (dart:convert/string_conversion.dart:300:5)
#3      _ConverterStreamEventSink.add (dart:convert/chunked_conversion.dart:69:18)
#4      _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:111:24)
#5      _RootZone.runUnaryGuarded (dart:async/zone.dart:1594:10)
#6      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:365:11)
#7      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:297:7)
#8      _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:784:19)
#9      _StreamController._add (dart:async/stream_controller.dart:658:7)
#10     _StreamController.add (dart:async/stream_controller.dart:606:5)
#11     _Socket._onData (dart:io-patch/socket_patch.dart:2455:41)
#12     _RootZone.runUnaryGuarded (dart:async/zone.dart:1594:10)
#13     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:365:11)
#14     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:297:7)
#15     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:784:19)
#16     _StreamController._add (dart:async/stream_controller.dart:658:7)
#17     _StreamController.add (dart:async/stream_controller.dart:606:5)
#18     new _RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:1942:33)
#19     _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:1385:14)
#20     _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
#21     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)
#22     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
#23     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:185:5)
```

**Comment by EvgenSilkov** [comment]
When updating patrol be careful that native instrumentation lib may get updates as well
Installation manual says to add this line into `/android/app/build.gradle`
     androidTestUtil "androidx.test:orchestrator:1.4.2"
after version bump it become
    androidTestUtil "androidx.test:orchestrator:1.5.1"
 
 so here is the correct update flow: 
 1 update patrol_cli
 2 bump patrol lib version in pubspec.yaml
 3 update native instrumentation if needed
 4 delete old instrumentation from target device, cause it may reject to install the new version

**Comment by kkabanova** [comment]
I also have this issue

Unhandled exception:
FormatException: Missing extension byte (at offset 1066)
#0      _Utf8Decoder.convertChunked (dart:convert-patch/convert_patch.dart:1959:7)
#1      _Utf8ConversionSink.addSlice (dart:convert/string_conversion.dart:310:28)
#2      _Utf8ConversionSink.add (dart:convert/string_conversion.dart:306:5)
#3      _ConverterStreamEventSink.add (dart:convert/chunked_conversion.dart:70:18)
#4      _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:115:24)
#5      _RootZone.runUnaryGuarded (dart:async/zone.dart:1778:10)
#6      _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:381:11)
#7      _BufferingStreamSubscription._add (dart:async/stream_impl.dart:312:7)
#8      _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:798:19)
#9      _StreamController._add (dart:async/stream_controller.dart:663:7)
#10     _StreamController.add (dart:async/stream_controller.dart:618:5)
#11     _Socket._onData (dart:io-patch/socket_patch.dart:2907:41)
#12     _RootZone.runUnaryGuarded (dart:async/zone.dart:1778:10)
#13     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:381:11)
#14     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:312:7)
#15     _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:798:19)
#16     _StreamController._add (dart:async/stream_controller.dart:663:7)
#17     _StreamController.add (dart:async/stream_controller.dart:618:5)
#18     new _RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:2326:31)
#19     _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:1[649](https://gitlab.gosystem.io/startups/client/umc/umc-templates-2.0/umc-template-new-jersey/-/jobs/15357118#L649):14)
#20     _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
#21     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)
#22     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:127:13)
#23     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:194:5)

Patrol CLI 3.6.0
Flutter 3.29.3

**Comment by EvgenSilkov** [comment]
It happens more in VS Code integrated terminal, Switching to Ghostty makes it happen less

**Comment by andriiShutov** [comment]
Also have this issue. How to fix this?

**Comment by dorklein** [comment]
Same here, I don't really understand why it's happening

**Comment by Noxiver** [comment]
I hit the same error and managed to fix it by combining online guides and an AI assistant.

This exception isn’t a terminal/locale issue. It happens when the response body that’s being printed is not valid UTF-8. Typical reasons:
	•	the server returns compressed bytes (gzip/br) but omits Content-Encoding,
	•	the body contains non-UTF-8 / binary data,
	•	logging tries to do utf8.decode(...) directly on those bytes.

The crash usually appears right after a log line like Response Text: because the UTF-8 decoder throws inside dart:convert before your code can catch it.

Minimal, safe fix (doesn’t change requests)

Keep your requests untouched. Only make response decoding lenient for logging.

1) Add a lenient transformer

Use SyncTransformer (or BackgroundTransformer) as the base and delegate transformRequest unchanged. Handle only text/json responses; for everything else fall back to the base transformer.
```
// log_decoder.dart
import 'dart:convert';
import 'dart:io';
import 'package:dio/dio.dart';

class LenientUtf8Transformer extends Transformer {
  final Transformer _base = SyncTransformer();

  @override
  Future<String> transformRequest(RequestOptions options) {
    // Do not modify outgoing requests.
    return _base.transformRequest(options);
  }

  @override
  Future transformResponse(RequestOptions options, ResponseBody response) async {
    final ct = (response.headers[Headers.contentTypeHeader] ?? const <String>[])
        .join(',')
        .toLowerCase();

    final expectsJson =
        options.responseType == ResponseType.json || ct.contains('application/json');
    final expectsText = ct.startsWith('text/');

    // Non-text/json -> use Dio's default behavior.
    if (!expectsJson && !expectsText) {
      return _base.transformResponse(options, response);
    }

    // Collect all bytes.
    final bytes = await response.stream.fold<List<int>>(<int>[], (a, b) => a..addAll(b));

    // Try UTF-8, then gzip+UTF-8, then UTF-8 with allowMalformed.
    String text;
    try {
      text = utf8.decode(bytes);
    } catch (_) {
      try {
        text = utf8.decode(gzip.decode(bytes));
      } catch (_) {
        text = utf8.decode(bytes, allowMalformed: true);
      }
    }

    if (expectsJson) {
      try {
        return jsonDecode(text);
      } catch (_) {
        // If the JSON is truly malformed, return the raw text so the logger can print it.
        return text;
      }
    }
    return text; // text/*
  }
}
```

2) Enable it only where you need it (before test)
```
final dios = getIt.getAll<Dio>();           // however you obtain your Dio instances
final transformer = LenientUtf8Transformer();

for (final dio in dios) {
  dio.transformer = transformer;            // requests remain unchanged
  dio.interceptors.removeWhere((i) => i is LogInterceptor);
  dio.interceptors.add(LogInterceptor(
    request: true,
    requestHeader: true,
    requestBody: true,
    responseHeader: true,
    responseBody: true,                     // keep printing the body
    error: true,
  ));
}
```
I hope this solution helps you too

**Comment by EvgenSilkov** [comment, marked as answer]
fix is in master, probably will be released in patrol_cli 3.11.0.
Meanwhile on 3.10.0
```bash
dart pub global deactivate patrol_cli 
git clone https://github.com/leancodepl/patrol.git
dart pub global activate --source path ./patrol/packages/patrol_cli
```
